File: analysis/test/data/codecharta/dependencyproject/src/main/java/de/sots/cellarsandcentaurs/adapter/persistence/PersistedCreatures.java
package de.sots.cellarsandcentaurs.adapter.persistence;

import de.sots.cellarsandcentaurs.domain.model.Creature;
import de.sots.cellarsandcentaurs.domain.model.CreatureId;
import de.sots.cellarsandcentaurs.domain.model.NoSuchCreatureException;
import de.sots.cellarsandcentaurs.domain.service.Creatures;

public class PersistedCreatures implements Creatures {

    private final CreatureRepository repository;

    public PersistedCreatures(CreatureRepository repository) {
        this.repository = repository;
    }

    @Override
    public void save(Creature creature) {
        repository.save(new CreatureEntity(creature.getId().id()));
    }

    @Override
    public Creature find(CreatureId id) throws NoSuchCreatureException {
        CreatureEntity creatureEntity = repository.findById(id.id()).orElseThrow(() -> new NoSuchCreatureException(id));
        return new Creature(new CreatureId(creatureEntity.getId()));
    }
}

File: training/ruby/lib/de/sots/cellars_and_centaurs/adapter/persistence/persisted_creatures.rb
# frozen_string_literal: true

require_relative '../../version'
require_relative 'creature_repository'
require_relative 'creature_entity'
require_relative '../../domain/service/creatures'
require_relative '../../domain/model/creature'
require_relative '../../domain/model/creature_id'
require_relative '../../domain/model/errors'
require_relative '../../application'

module De::Sots::CellarsAndCentaurs::Adapter::Persistence
  class PersistedCreatures
    include De::Sots::CellarsAndCentaurs::Domain::Service::Creatures

    Model = De::Sots::CellarsAndCentaurs::Domain::Model

    def initialize(repository = CreatureRepository.new)
      @repository = repository
    end

    def save(creature)
      @repository.save(CreatureEntity.new(creature.id.id, creature.type))
    end

    def find(creature_id)
      creature_entity = @repository.find_one(creature_id.id)
      raise Model::NoSuchCreatureException, creature_id if creature_entity.nil?

      to_domain(creature_entity)
    end

    private

    def to_domain(creature_entity)
      type = creature_entity.type || De::Sots::CellarsAndCentaurs::Application::CreatureFacade::STANDARD_CREATURE_TYPE
      Model::Creature.new(Model::CreatureId.new(creature_entity.id), type)
    end
  end
end
